node/pkg/watchers/hedera/watcher.go

```diff
@@ -0,0 +1,612 @@
+	// 	Sender           common.Address
+	// 	Sequence         uint64
+	// 	Nonce            uint32
+	// 	Payload          []byte
+	// 	ConsistencyLevel uint8
+	// 	Raw              types.Log // Blockchain specific contextual infos
+	// }
+
+	seq := unpackedMsg[0].(uint64)
+	nonce := unpackedMsg[1].(uint32)
+	payload := unpackedMsg[2].([]byte)
+	cLevel := unpackedMsg[3].(uint8)
+	logger.Info("unpackedMsg",
+		zap.Int("Length of unpackedMsg", len(unpackedMsg)),
+		zap.Uint64("Sequence", seq),
+		zap.Uint32("Nonce", nonce),
+		zap.Uint8("consistencyLevel", cLevel))
+	var chainID vaa.ChainID = 27
+
+	// Convert timestamp
+	timeVals := strings.Split(timeStampBase.String(), ".")
+	timeValSeconds, err := strconv.ParseInt(timeVals[0], 10, 64)
+	if err != nil {
+		logger.Error("Failed to convert seconds", zap.String("sec", timeVals[0]))
+		return nil, fmt.Errorf("Failed to convert time seconds %s", timeVals[0])
+	}
+	timeValNS, err := strconv.ParseInt(timeVals[1], 10, 64)
+	if err != nil {
+		logger.Error("Failed to convert nanoseconds", zap.String("nanosec", timeVals[1]))
+		return nil, fmt.Errorf("Failed to convert time nanoseconds %s", timeVals[1])
+	}
+	unixTS := time.Unix(timeValSeconds, timeValNS)
+	logger.Info("timestamps", zap.Stringer("base", timeStampBase), zap.String("slice0", timeVals[0]), zap.String("slice1", timeVals[1]), zap.Stringer("unix", unixTS))
+
+	messagePublication := &common.MessagePublication{
+		TxHash:           txHash,      // In log
+		Timestamp:        unixTS,      // In log
+		Nonce:            nonce,       // In log data
+		Sequence:         seq,         // In log data
+		EmitterChain:     chainID,     // Don't know where to get this
+		EmitterAddress:   emitterAddr, // In log
+		Payload:          payload,     // In log data
+		ConsistencyLevel: cLevel,      // In log data
+	}
+
+	logger.Info("messagePublication",
+		zap.Stringer("txHash", txHash),
+		zap.Stringer("timestamp", unixTS),
+		zap.Uint32("nonce", nonce),
+		zap.Uint64("sequence", seq),
+		zap.Stringer("EmitterAddr", emitterAddr),
+		zap.Uint8("ConsistenceLevel", cLevel),
+	)
+
+	return messagePublication, nil
+}
+
+// TxMessageToEvent a JSON structured "contracts" output and converts it into an event
+func TxMessageToEvent(logger *zap.Logger, msg string) ([]*common.MessagePublication, error) {
+
+	ethAbi, err := abi2.JSON(strings.NewReader(ethabi.AbiABI))
+	if err != nil {
+		logger.Fatal("failed to parse Eth ABI", zap.Error(err))
+		return nil, err
+	}
+
+	// Check existence of required values in outer scope
+	txHashBase := gjson.Get(msg, "hash")
+	if !txHashBase.Exists() {
+		return nil, errors.New("Message has no transaction hash")
+	}
+	txhash := txHashBase.String()
+	txHash := eth_common.HexToHash(txhash)
+
+	timeStampBase := gjson.Get(msg, "timestamp")
+	if !timeStampBase.Exists() {
+		return nil, fmt.Errorf("Message has no timestamp field for txhash %s", txhash)
+	}
+
+	// Convert timestamp
+	timeVals := strings.Split(timeStampBase.String(), ".")
```

Comment:
You're doing this twice. Maybe create a function?

Reply:
Done